### ISSUE TITLE
Type priacy of diagm()

### ISSUE THREAD

**Issue by oxinabox** [JuliaArrays/StaticArrays.jl#961]
Found when resolving https://github.com/JuliaDiff/ChainRulesCore.jl/issues/472
where I was like "I fixed the type-piracy, why is it still not saying that the module is from LinearAlgebra.
I guess this one never happens in the wild because 0 argument `diagm` is not really useful but:

```julia
julia> using LinearAlgebra

julia> diagm()
0×0 BitMatrix
```

Add StaticArrays into the mix and
```julia
julia> using LinearAlgebra, StaticArrays

julia> diagm()
ERROR: ArgumentError: reducing over an empty collection is not allowed
Stacktrace: ...

julia> @which diagm()
diagm(kvs::Pair{var"#s43", var"#s42"} where {var"#s43"<:Val, var"#s42"<:(StaticVector{N, T} where {N, T})}...) in StaticArrays at /home/oxinabox/.julia/packages/StaticArrays/vxjOO/src/linalg.jl:142
```

This is a motivating case for JuliaLang/julia#42222

**Comment by mateuszbaran** [comment]
Good catch, PR #962 should fix this.

**Comment by mateuszbaran** [comment]
Solved by #962
